Mqtt abstraction layer for subscribing to node event api's

rebinmannady93 opened the pull request:
# Description of change

fixes #169.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

TBD

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

rebinmannady93 committed:
Mqtt wrapper initial files

rebinmannady93 committed:
Adding MQTT_CLIENT_ENABLE flag in cmake

rebinmannady93 committed:
Removing unwanted line

rebinmannady93 committed:
Cleared a bug. response.u.received_milestone_latest->index --> response->u.received_milestone_latest->index

rebinmannady93 committed:
Added mosquitto.cmake

rebinmannady93 committed:
Handling error cases, free all dynamic objects.

Signed-off-by: Muhammed Rabin <rabin.muhammed@iota.org>

rebinmannady93 committed:
Added support for subscribing multiple topics.

Signed-off-by: Muhammed Rabin <rabin.muhammed@iota.org>

oopsmonk commented on CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ option(IOTA_ASAN_ENABLED "Enable Address sanitizer" OFF)
 option(ENABLED_JEMALLOC "Enable jemalloc memory allocator" OFF)
 option(IOTA_EXAMPLES "Build example application" ON)
 option(IOTA_WALLET_ENABLE "Enable simple wallet API" ON)
+option(MQTT_CLIENT_ENABLE "Enable simple wallet API" OFF)
```
```suggestion
option(MQTT_CLIENT_ENABLE "Enable MQTT Event APIs" OFF)
```

rebinmannady93 replied:
Done.

oopsmonk commented on cmake/mosquitto.cmake:
```diff
@@ -0,0 +1,36 @@
+#[[
+// Copyright 2020 IOTA Stiftung
```
```suggestion
// Copyright 2021 IOTA Stiftung
```

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/api/mqtt/sub_milestone_latest.c:
```diff
@@ -0,0 +1,79 @@
+
+// Copyright 2020 IOTA Stiftung
```
```suggestion
// Copyright 2021 IOTA Stiftung
```

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/api/mqtt/sub_milestone_latest.c:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  // Allocates milestone latest response object
+  res_milestone_latest_t *res = res_milestone_latest_new();
```
checks if the object is NULL.

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/api/mqtt/sub_milestone_latest.c:
```diff
@@ -0,0 +1,79 @@
+  // Parse index
+  if ((json_get_uint32(json_obj, JSON_KEY_INDEX, &(res->u.received_milestone_latest->index))) != 0) {
+    printf("[%s:%d]: parse %s failed\n", __func__, __LINE__, JSON_KEY_INDEX);
+    res->is_error = true;
+    char *error_msg = "Parse index failed";
+    res->u.error = (char*)malloc(strlen(error_msg));
+    strcpy(res->u.error, error_msg);
```
it's good that return an error object but it should prevent dynamic allocations. since there is an error message via `printf`, it seems redundant.

rebinmannady93 replied:
In this case, 
Option 1 : Inseatd of using res_milestone_latest_t structure
as paramter for callback function, I can directly use milestone_latest_t as the parameter for callback. I used res_milestone_latest_t structure format to keep the http response and mqtt response similar to each other. (Note : But here bool is_error field wont be present, we will pass data to user defined callback only if received response is as expected)

Option 2 :  Change res_milestone_latest_t  to
typedef struct {
  bool is_error;  ///< True if error occured
  milestone_latest_t *received_milestone_latest;  ///< a balance object if is_error is False
} res_milestone_latest_t;

Which do you suggest?

rebinmannady93 replied:
Made Changes.

oopsmonk commented on src/client/api/mqtt/sub_milestone_latest.c:
```diff
@@ -0,0 +1,79 @@
+  // Delete json object created
+  cJSON_Delete(json_obj);
+  // Free response object
+  res_milestone_latest_free(res);
```
the `res` will be used by a callback and might be used after freeing.

rebinmannady93 replied:
res_milestone_latest_new gets called and the memory gets dynamically initialzes evertime a message is recevied in the subscribed topic. My idea is that from the callback function, the received data needs to be copied to other variable instead of using the same variable out of scope of the callback function. If we don't free it then for every message received, memory will get allocated. In the current case, once a message is received, buffer memory get allocated internally to store the data, it is passed to the callback function and after the callback gets executed, the internally allocated memory is freed.

Or should we not free the variable, and write a comment at the callback function method specifying to free the memory on every message received?

rebinmannady93 replied:
Made changes.

oopsmonk commented on src/client/api/mqtt/sub_milestone_latest.h:
```diff
@@ -0,0 +1,57 @@
+// Copyright 2020 IOTA Stiftung
```
```suggestion
// Copyright 2021 IOTA Stiftung
```

rebinmannady93 replied:
Done.

rebinmannady93 committed:
Clang Format Fix.

Signed-off-by: Muhammed Rabin <rabin.muhammed@iota.org>

rebinmannady93 committed:
Some typo corrections.

Signed-off-by: Muhammed Rabin <rabin.muhammed@iota.org>

rebinmannady93 committed:
Checking initialised object is null.
Signed-off-by: Muhammed Rabin <rabin.muhammed@iota.org>

rebinmannady93 committed:
1. Removed usage of golbal mosquitto instance mqtt_mosquitto.c
2. Passing client instance as userdata so it can be used in callback functions, this helps to avoid use of global mosquitto instance.
3. Accept qos as a subscription function parameter.

rebinmannady93 committed:
Added node_event layer that act as a wrapper to mqtt network layer.

rebinmannady93 committed:
Added Function Callbacks and Node Event API's

rebinmannady93 committed:
Added parser for milestone/latest topic.

oopsmonk commented on examples/get_event_messages.c:
```diff
@@ -0,0 +1,63 @@
+      event_subscribe(event->client, NULL, "milestones/latest", 1);
+      break;
+    case NODE_EVENT_DISCONNECTED:
+      printf("Node event vetwork disonnected\n");
```
typo
```suggestion
      printf("Node event network disconnected\n");
```

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/api/events/sub_milestone_latest.c:
```diff
@@ -0,0 +1,42 @@
+
+// Copyright 2021 IOTA Stiftung
```
```suggestion
// Copyright 2021 IOTA Stiftung
```

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/network/mqtt/mqtt_mosquitto.c:
```diff
@@ -0,0 +1,196 @@
+#include <string.h>
+#include "client/network/mqtt/mqtt.h"
+
+struct mqtt_client {
+  struct mosquitto *mosq;
+  mqtt_client_config_t *config;
+  void (*mqtt_callback)(mqtt_client_event_t *event, void *userdata);
+  void *cb_userdata;
+};
```
I think it should be declared in `client/network/mqtt/mqtt.h`

rebinmannady93 replied:
My idea is that, struct mqtt_client will depend on the mqtt client. 
For eg now since we use mosquitto as mqtt client for linux we have struct mosquitto *mosq declared in struct mqtt_client in file mqtt_mosquitto.c.
For esp32 similarly esp_mqtt_client_handle_t client will be declared in struct mqtt_client in file mqtt_esp32.c.

oopsmonk commented on src/client/network/mqtt/mqtt_mosquitto.c:
```diff
@@ -0,0 +1,196 @@
+  mosquitto_destroy(client->mosq);
+  // Call to free resources associated with the library.
+  mosquitto_lib_cleanup();
+  // Free client insatnce
```
typo
```suggestion
  // Free client instance
```

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/api/events/sub_milestone_latest.c:
```diff
@@ -0,0 +1,42 @@
+  milestone_latest_t *res = (milestone_latest_t *)malloc(sizeof(milestone_latest_t));
+  if (res == NULL) {
+    printf("[%s:%d] OOM\n", __func__, __LINE__);
+    return;
```
NULL or `res` should be passed to the caller.
```suggestion
    return NULL;
```

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/network/mqtt/mqtt_mosquitto.c:
```diff
@@ -0,0 +1,196 @@
+
+/* Callback called when the client receives a CONNACK message from the broker. */
+void on_mqtt_connect(struct mosquitto *mosq, void *client, int reason_code) {
+  int rc;
```
unused variable.

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/network/mqtt/mqtt_mosquitto.c:
```diff
@@ -0,0 +1,196 @@
+    mqtt_event->event_id = MQTT_CONNECTED;
+  } else {
+    mqtt_event->event_id = MQTT_ERROR;
+    char *con_ack = mosquitto_connack_string(reason_code);
```
please fix ‘const’ qualifier warnings, it means `data` element is the `char const *` in mqtt_client_event_t

rebinmannady93 replied:
Done.

oopsmonk reviewed (changes requested):
typos and a potential issue need to be fixed, thanks.

rebinmannady93 committed:
Fixed typos and few potential errors.

rebinmannady93 committed:
return; -> return NULL;

oopsmonk commented on examples/get_event_messages.c:
```diff
@@ -0,0 +1,63 @@
+      event_subscribe(event->client, NULL, "milestones/latest", 1);
+      break;
+    case NODE_EVENT_DISCONNECTED:
+      printf("Node event network disonnected\n");
```
```suggestion
      printf("Node event network disconnected\n");
```

rebinmannady93 replied:
Done.

oopsmonk reviewed (changes requested):
LGFM, please fix the typo.

rebinmannady93 committed:
Typo Fix.
